### PR TITLE
Registration API

### PR DESCRIPTION
**Pull request by juliushaertl**
This PR implements a basic registration API as an OCS endpoint as discussed in #41 

I've included the commit from #76 here, since a lot of refactoring has been done based on this and I think it might easier to review the changes together.

ToDo:
- [x] Write API documentation
- [x] Unit tests
- [x] The API should just use client secret as identifier for the registration, because otherwise users could create accounts by verifying with the token without receiving an email

Example usage is documented here for now with some basic curl commands:

https://gist.github.com/juliushaertl/5a1d1132e7370b5ad38fbd6da3cae5b8#example-usage

@pellaeon @rullzer 

Maybe @pierreozoux @Gomez want to have a look as well. :wink:

### CODE REVIEW

File: capabilities.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ *  along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Registration;
+
+use OCP\Capabilities\ICapability;
+use OCP\IURLGenerator;
+
+class Capabilities implements ICapability {
```
Comment (rullzer) [inline capabilities.php]: Capabilties can only be queried (currently) when authenticated. https://github.com/nextcloud/server/issues/4510

File: capabilities.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	public function __construct(IURLGenerator $urlGenerator) {
+		$this->urlGenerator = $urlGenerator;
+	}
+
+	public function getCapabilities() {
+		return [
+			'registration' =>
+			[
+				'enabled' => true,
+				'apiRoot' => $this->urlGenerator->linkTo(
```
Comment (rullzer) [inline capabilities.php]: Please use v2.php

File: controller/apicontroller.php [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	/** @var MailService */
+	private $mailService;
+	/** @var IL10N */
+	private $l10n;
+	/** @var Defaults */
+	private $defaults;
+
+	public function __construct($appName,
+								IRequest $request,
+								$corsMethods = 'PUT, POST, GET, DELETE, PATCH',
```
Comment (rullzer) [inline controller/apicontroller.php]: kill this it is the same as the parent

File: controller/apicontroller.php [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+								$corsAllowedHeaders = 'Authorization, Content-Type, Accept',
```
Comment (rullzer) [inline controller/apicontroller.php]: same

File: controller/apicontroller.php [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+								$corsMaxAge = 1728000,
```
Comment (rullzer) [inline controller/apicontroller.php]: same

File: controller/apicontroller.php [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+								RegistrationService $registrationService,
+								MailService $mailService,
+								IL10N $l10n,
+								Defaults $defaults) {
+		parent::__construct($appName, $request, $corsMethods, $corsAllowedHeaders, $corsMaxAge);
```
Comment (rullzer) [inline controller/apicontroller.php]: Remove the args with defaults here if they are the same ;)

File: controller/apicontroller.php [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+		$this->mailService = $mailService;
+		$this->l10n = $l10n;
+		$this->defaults = $defaults;
+	}
+
+	/**
+	 * @PublicPage
+	 * @AnonRateThrottle(limit=5, period=1)
+	 *
+	 * @param $username
```
Comment (rullzer) [inline controller/apicontroller.php]: use the typehints ;)

```php
@param type $var
```

So

```php
@param string $username
```

File: service/registrationexception.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+class RegistrationException extends \Exception {
+
+	protected $hint;
+
+	public function __construct($message, $hint = "") {
+		parent::__construct($message);
+		$this->setHint($hint);
+	}
+
+	public function setHint($hint) {
```
Comment (rullzer) [inline service/registrationexception.php]: Setter for exception seems weird....

**Comment by pellaeon** [comment]
Hi @juliushaertl I looked over the changes and most looked good, and with the design of "pending" status it should be feasible if we want to incorporate admina approval feature in the future.

Some notes while I was reading the code:

### Client secret
Client secret is held only by the client app, and is used to uniquely identify the client app making the registration request.

Q: Why is there need for a client secret at all? why not just check the registration status by Token?
A: a Token is only received when the email successfully delivers, and it's not received by the app, so there's no way for the client app to check registration status other than introducing client secret

### Different behavior of `verifyToken` when registering with OCS API vs normal web form
When registering through normal web form, Token is verified **before** username and password are provided by the user (but the email is already provided).
When registering through OCS API, Token is verified **after** username and password are provided.
So the new `verifyToken` controller checks if username and password is already provided, if so, a "successfully registered" message is printed, if not, the old form is presented, and the user fills in username and password.

Some of the questions I haven't checked, will do that when I have more time, or you can just answer them ;-)
- how does the new OCS flow handles resending verification tokens?
- how does the new OCS flow handles things like "username squatting"? eg. an attacker can repeatedly register with popular usernames while not verifying the email, will new users be able to register if the username is already taken but the email hasn't been verified? or is there some kind of expiration mechanism?

I may think of more questions, I'll post them when I do.

In the mean time, I hope you start writing the documentations :-)

Great work! 🚀

**Comment by juliushaertl** [comment]
@pellaeon Thanks for your feedback. I really appreciate it. 

> Q: Why is there need for a client secret at all? why not just check the registration status by Token?
A: a Token is only received when the email successfully delivers, and it's not received by the app, so there's no way for the client app to check registration status other than introducing client secret

Exactly. The token should not be exposed anywhere else than in the email. Otherwise that would allow users to verify their address without receiving an email.

> how does the new OCS flow handles resending verification tokens?

A new token will be generated and sent to the users email. See https://github.com/pellaeon/registration/pull/77/files#diff-b9e15819672f6817a033ecc447a6e2a2R153

> how does the new OCS flow handles things like "username squatting"? eg. an attacker can repeatedly register with popular usernames while not verifying the email, will new users be able to register if the username is already taken but the email hasn't been verified? or is there some kind of expiration mechanism?

I have not thought about that kind of attack vector until now. But I guess we could at least add the AnonRateThrottle rate limit annotation that Nextcloud has introduced here. At the moment there is no check if there already is a pending registration for the username, but i'll add that as well. I need to think a bit more about this, maybe we need some kind of expiration, as you said.

I'll try to finish documentation of the API and the unit tests later today.

**Comment by juliushaertl** [comment]
Sorry for the delay. I've added unit tests at least for the new code parts. @pellaeon It might make sense to enable travis ci or some similar ci service on the repo, so we can see if some patches break the unit tests in the future.

The API documentation can be found here: https://gist.github.com/juliushaertl/5a1d1132e7370b5ad38fbd6da3cae5b8

**Comment by pellaeon** [comment]
Hey @juliushaertl , I encountered this error while upgrading the plugin:
```
Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'ALTER TABLE oc_registration ADD COLUMN "username" VARCHAR(255) NOT NULL': SQLSTATE[HY000]: General error: 1 Cannot add a NOT NULL column with default value NULL
```

Might be some problems with my Doctrine or MariaDB, I'm looking into it, please let me know if you have a hint.

**Comment by pellaeon** [comment]
Oops, I forgot I had sqlite3 instead of MariaDB on my test server.
Since sqlite3 is only used for testing purposes, I think the user may just drop the existing table and re-enable the plugin.

**Pull request merged**